endpoint: store more contextual information about datapath regenerations while regenerating endpoints

ianvernon opened the pull request:
This is a revival / update of #5616. 

It makes `regenerationContext` private to the endpoint package, and ensures that a given `regenerationContext` is created in a single goroutine for a given endpoint so as to not have multiple threads accidentally modifying the same structure concurrently. 

Information that is related to the state of the datapath for a regeneration is stored in a new member of `regenerationContext`, called `datapathRegenerationContext`. I'm open to changing the name of these types if desired. Storing information in this type allows for `regenerateBPF` to be split up into separate functions, which consume this type. It also ensures that these functions can defer the unlocking of the endpoint if needed in case there is a panic that arises. This reduces the risk of deadlock, as there are other functions in `regenerateBPF` which are deferred and lock the endpoint structure, which would deadlock the agent if a panic occurred while `regenerateBPF` was holding the endpoint lock.

This change makes `regenerateBPF` easier to divide up into logical, more readable chunks. It also ensures that state related to regeneration is stored in a type, which we can remove / add members from if so desired, while maintaining the same function signatures. 

I also believe that grouping more of the information related to plumbing the datapath into a single type will help us figure out how we can hide `pkg/bpf` from the endpoint package in the future. 

I tried to make this as easy to review as possible by having a lot of small, easy to digest commits.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6191)
<!-- Reviewable:end -->

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -459,34 +459,53 @@ func (e *Endpoint) removeOldRedirects(owner Owner, desiredRedirects map[string]b
-	)
+// datapathRegenerationContext contains information related to regenerating the
+// datapath (BPF, proxy, etc.).
+type datapathRegenerationContext struct {
```
I wonder if this structure should be part of the RegenerationContext itself. Open to suggestions.

ianvernon replied:
Also am open to different names for these fields, etc.

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -655,87 +662,134 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
 	stats.prepareBuild.End(true)
-	if bpfHeaderfilesChanged || regenContext.ReloadDatapath {
-		closeChan := loadinfo.LogPeriodicSystemLoad(log.WithFields(logrus.Fields{logfields.EndpointID: epID}).Debugf, time.Second)
+	if bpfRegenContext.bpfHeaderfilesChanged || regenContext.ReloadDatapath {
```
Maybe ReloadDatapath should be moved into `datapathRegenerationContext` - thoughts ?

raybejjani replied:
That's reasonable, I think. We can define the inner struct directly since it won't ever be referred to by type.

ianvernon replied:
Fixed in latest upload.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -459,34 +459,53 @@ func (e *Endpoint) removeOldRedirects(owner Owner, desiredRedirects map[string]b
 	}
 
-	epID := e.StringID()
+	// defer Unlock so that it will be done even if something panics
```
remove this comment. We shouldn't care about panics at the point on adding comments before doing a `defer e.Unlock()`.

ianvernon replied:
Done.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -459,34 +459,53 @@ func (e *Endpoint) removeOldRedirects(owner Owner, desiredRedirects map[string]b
-	epID := e.StringID()
+	// defer Unlock so that it will be done even if something panics
+	defer e.Unlock()
+	stats.waitingForLock.End(err == nil)
```
1) There's no `stats.waitingForLock.Start()`
2) This should be between line 499 and 500

ianvernon replied:
Will be fixed in next upload.

ianvernon replied:
Fixed in latest upload.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -655,87 +662,134 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
+
+	// Note that revFunc may hold the endpoint lock.
+	if revFunc != nil {
+		defer revFunc(reterr)
```
reterr will always be nil so the revFunc won't do the right things if `reterr != nil`

raybejjani replied:
This is because the value of retErr is resolved when the defer is made (vs executed). To have it capture retErr later `defer func() { revFunc(retErr) }()` should do it (fmt may force it to be multiline, however)

ianvernon replied:
Fixed in latest upload. Thanks for catching this guys, I did not realize this subtlety.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -655,87 +662,134 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
+
+	// Note that revFunc may hold the endpoint lock.
+	if revFunc != nil {
+		defer revFunc(reterr)
```
the `revFunc` also does a `Lock()`, if a panic occurs, let's say on line 816, the same bug will happen.

ianvernon replied:
:( 

This is true. However, the executed code path is fairly small,  which we have audited to avoid panics.

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -459,34 +459,53 @@ func (e *Endpoint) removeOldRedirects(owner Owner, desiredRedirects map[string]b
+	epInfoCache           *epInfoCache
+	bpfHeaderfilesChanged bool
+	proxyWaitGroup        *completion.WaitGroup
+	ctCleaned             chan struct{}
```
I think I have a slight preference to returning this channel from `runPreCompilationSteps`. I'm not sure that it's more clear but we seem to use it almost immediately after `runPreCompilationSteps` is called. I would go further and do the wait inside `runPreCompilationSteps` but we seem to have a drymode check (but perhaps we can check that and close the channel on create, like the else clause on line 568.
Separately, do you know why we set `ctCleaned=true` so late in `regenerateBPF` (line 822)? It seems like we could set it right when we stop waiting for the clean goroutine.

ianvernon replied:
The attractive thing about storing this field within the `datapathRegenerationContext` is that we can refactor it as needed instead of having a large amount of state included in a single function. In the future I'd like to break up `runPreCompilationSteps` to be smaller and more modular if possible.

Regarding your last question, I'm not sure. I'd have to dig through the `git log`.

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -459,34 +459,53 @@ func (e *Endpoint) removeOldRedirects(owner Owner, desiredRedirects map[string]b
+	}
 
 	// In the first ever regeneration of the endpoint, the conntrack table
 	// is cleaned from the new endpoint IPs as it is guaranteed that any
```
Absolutely out of scope of this PR but I wonder if the ctCleaned check can become a `sync.Once` one day.

ianvernon replied:
Good point.

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -501,69 +520,59 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
-		if reterr == nil {
+
+	proxyCleanupFunc := func(err error) {
+		if err == nil {
```
Should we also defer the unlocks in this function? Sadly, one is an `UnconditionalLock` and the other a `LockAlive`

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -582,18 +591,17 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
-			return 0, compilationExecuted, fmt.Errorf("unable to regenerate policy for '%s': %s", e.PolicyMap.String(), err)
+			return proxyCleanupFunc, fmt.Errorf("unable to regenerate policy for '%s': %s", e.PolicyMap.String(), err)
 		}
+		stats.policyCalculation.End(err == nil)
```
I'm surprised this moved. The earlier logic seemed to call end irrespective of the `err != nil` return.

ianvernon replied:
Will be fixed in next upload.

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -459,34 +459,53 @@ func (e *Endpoint) removeOldRedirects(owner Owner, desiredRedirects map[string]b
+	// defer Unlock so that it will be done even if something panics
+	defer e.Unlock()
+	stats.waitingForLock.End(err == nil)
+	if err = datapathRegenCtx.completionCtx.Err(); err != nil {
```
Can this happen anymore? we make the context and almost immediately enter `runPreCompilationSteps` without anything in between that would cause the error. Is this only in case the lock wait was longer than the context timeout?

ianvernon replied:
Most likely? You'd have to confirm with @jrajahalme , as I did not change when this logic was invoked here.

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -459,34 +459,53 @@ func (e *Endpoint) removeOldRedirects(owner Owner, desiredRedirects map[string]b
+	ctCleaned             chan struct{}
+	completionCtx         context.Context
+	completionCancel      context.CancelFunc
+}
```
I'm surprised `currentDir` and `nextDir` aren't in here too. We're passing them about anyway and they seem relevant to the context? (maybe not?)

ianvernon replied:
@raybejjani I made this structure contain values that were only local to `regenerateBPF` note that `currentDir` and `nextDir` are passed as parameters to `regenerateBPF`. I can add these to the `datapathRegenerationContext` if I store that information within the `RegenerationContext` structure itself. I'll do that as part of my next upload.

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -655,87 +662,134 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
+// boolean if the BPF compilation was executed and an error in case of an error.
+func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenContext *RegenerationContext) (revnum uint64, compiled bool, reterr error) {
+	stats := &regenContext.Stats
+	stats.waitingForLock.Start()
```
This might read better if it was right before `GetCompilationLock`. I also noticed that the matching end for it is in `runPrecompileSteps`. I think it's fine to instead `Start` and `End` around `GetCompilationLock` and `UnconditionalRLock` in this function, and the first `LockAlive` in `runPrecompileSteps`. The current measurement spans the two functions in an odd way, and doing it per lock here and there shouldn't break anything but keep the logic very clearly local.

ianvernon replied:
Will have it fixed in next upload.

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -655,87 +662,134 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
+	revFunc, err := e.runPreCompilationSteps(owner, bpfRegenContext, currentDir, nextDir, regenContext)
+
+	// Note that revFunc may hold the endpoint lock.
+	if revFunc != nil {
```
I wonder if it would be safer to make the contract of `runPreCompilationSteps` always return a non-nil function, but return a dummy on error.

ianvernon replied:
Only thing I can argue against that is that we have to perform an unnecessary function call, but I do like protecting against nil-dereferencing (which would cause a panic ;) ).

ianvernon replied:
I've changed this to be a function which is performed upon endpoint with two parameters, a `RegenerationContext` and an `error`. Let me know if you prefer how I've done it.

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -655,87 +662,134 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
+	// Returns an error, as well as a function which may need deferring
+	// based once this function returns. This function is intentionally not
+	// deferred in this function itself because it may acquire the endpoint
+	// lock.
```
Is the last sentence correct? I thought `revFunc` needs to be returned from `runCompilationSteps` and deferred in `regenerateBPF` because it only reverted if there was an error in either `runCompilationSteps` or `regenerateBPF` (the lock stuff also applies).
With that in mind perhaps it would be simpler to not check the error in `revFunc` but run it in `runCompilationSteps` if that returns an error, and separately run it here if regenerateBPF encounters an error. This might be worse, though, since we will likely do a `defer func() { if err != nil { proxyCleanup/revFunc() } }()` in each function anyway and that's a little more obtuse.

ianvernon replied:
My wording is confusing. By 'this function', I meant 'runPreCompilationSteps'. You interpreted it as 'regenerateBPF'. I will clarify the comment.

ianvernon replied:
I've removed this comment / returning a function to defer in the caller.

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -604,21 +612,20 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
+		err, networkPolicyRevertFunc = e.updateNetworkPolicy(owner, datapathRegenCtx.proxyWaitGroup)
 		stats.proxyPolicyCalculation.End(err == nil)
 		if err != nil {
 			e.Unlock()
```
This should be removed, right? the deferred one will do the right thing. There are a few more below as well.

ianvernon replied:
The deferred function performs any cleanup or finalization for the revert / finalize stacks, not actual plumbing of the network policy for the proxy.

raybejjani replied:
I don't fully remember why I asked this, but I remember thinking there was a deferred unlock. Looking again, I must have missed the lock calls in the big `reterr == nil` defer above.

raybejjani reviewed (dismissed):
I think I'll look this over again but I've made a bunch of comments. In particular, we might have left some now-uneeded unlocks in there.

Commit:
endpoint: move RegenerationContext to separate file

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: add datapathRegenerationContext type

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: move ReloadDatapath to datapathRegenerationContext in RegenerationContext

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: move ctCleaned in regenerateBPF into datapathRegenerationContext

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: move bpfHeaderfilesChanged in regenerateBPF into datapathRegenerationContext

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: move bpfHeaderfilesHash in regenerateBPF into datapathRegenerationContext

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: move epInfoCache in regenerateBPF into datapathRegenerationContext

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: move datapathRegenerationContext setup after locks are acquired

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: move wait group, completions into datapathRegenerationContext

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: move revertStack / finalizeList into datapathRegenerationContext in regenerateBPF

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: factor out deferred proxy finalizing / reverting logic into separate function

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: move currentDir / nextDir into datapathRegenerationContext in regenerateBPF

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: factor out code which locks endpoint before compilation into function

In this function, we can defer the unlocking of the endpoint's mutex.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: factor out function which holds endpoint lock within regenerateBPF into separate function

This makes regenerateBPF smaller and the code easier to understand. Also have
prepare build stat not overlap with waitingForCTClean stat, and End this
spanstat when said function returns.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: factor out compilation of BPF out of regenerateBPF into separate function

Signed-off by: Ian Vernon <ian@cilium.io>

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -726,7 +725,7 @@ func (e *Endpoint) regenerateBPF(owner Owner, regenContext *RegenerationContext)
 	stats.prepareBuild.End(true)
 	if datapathRegenCtxt.bpfHeaderfilesChanged || datapathRegenCtxt.reloadDatapath {
-		closeChan := loadinfo.LogPeriodicSystemLoad(log.WithFields(logrus.Fields{logfields.EndpointID: epID}).Debugf, time.Second)
+		closeChan := loadinfo.LogPeriodicSystemLoad(log.WithFields(logrus.Fields{logfields.EndpointID: e.StringID()}).Debugf, time.Second)
```
This accesses the endpoint's ID, but AFAICT the ID for an endpoint should never change during its lifetime (i.e., it is immutable).

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -822,7 +647,183 @@ func (e *Endpoint) realizeBPFState(regenContext *RegenerationContext) (compilati
+	stats.proxyConfiguration.End(true)
+
+	stats.prepareBuild.Start()
+	defer func() {
```
This is the only change in the code - I moved `prepareBuild.End` to be within this function instead of included with `waitingForCTClean`.

Commit:
endpoint: make RegenerationRequest private

Instead, consumers of pkg/endpoint pass down an ExternalRegenerationMetadata
structure, which contains data relevant to a regeneration which are populated
by said consumers (e.g., daemon).

When an endpoint regeneration is triggered, transfer the data from this new
type into a regenerationRequest. This regenerationRequest is only available
to the goroutine which is responsible for a single endpoint's regeneration.

Signed-off by: Ian Vernon <ian@cilium.io>

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -527,20 +526,162 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
+// runPreCompilationSteps runs all of the regeneration steps that are necessary
+// right before compiling the BPF for the given endpoint.
+// The endpoint mutex must not be held.
+// Returns a function which is invoked by the caller to revert some of the
```
Seems to be returning an `error` instead?

ianvernon replied:
Updated documentation in next upload.

jrajahalme commented on pkg/endpoint/regenerationcontext.go:
```diff
@@ -0,0 +1,22 @@
+// the endpoint subsystem should be made aware of for a given endpoint.
+type RegenerationRequest struct {
+	Reason         string
+	ReloadDatapath string
```
Why not `bool` here?

jrajahalme replied:
Seems this commit is replaced by the next, maybe squash these two together?

ianvernon replied:
Will squash, thanks!

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -526,234 +526,240 @@ func (e *Endpoint) regenerateBPF(owner Owner, regenContext *regenerationContext)
+		if e.bpfConfigMap == nil {
+			e.bpfConfigMap, _, err = bpfconfig.OpenMapWithName(e.BPFConfigMapPath(), e.BPFConfigMapName())
+			if err != nil {
+				e.Unlock()
```
Should not unlock here, already deferred!

ianvernon replied:
Good catch, fixed in next upload.

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -526,234 +526,240 @@ func (e *Endpoint) regenerateBPF(owner Owner, regenContext *regenerationContext)
+			// ELF file, but there's no solution to this just yet.
+			if err = e.bpfConfigMap.Update(e.desiredBPFConfig); err != nil {
+				e.getLogger().WithError(err).Error("unable to update BPF config map")
+				e.Unlock()
```
Should not unlock here either!

ianvernon replied:
Good catch, fixed in next upload.

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -871,7 +666,213 @@ func (e *Endpoint) realizeBPFState(regenContext *regenerationContext) (compilati
+	if e.bpfConfigMap == nil {
+		e.bpfConfigMap, _, err = bpfconfig.OpenMapWithName(e.BPFConfigMapPath(), e.BPFConfigMapName())
+		if err != nil {
+			e.Unlock()
```
SHould not unlock here!

ianvernon replied:
Good catch, will be fixed in next upload.

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -871,7 +666,213 @@ func (e *Endpoint) realizeBPFState(regenContext *regenerationContext) (compilati
+		// ELF file, but there's no solution to this just yet.
+		if err = e.bpfConfigMap.Update(e.desiredBPFConfig); err != nil {
+			e.getLogger().WithError(err).Error("unable to update BPF config map")
+			e.Unlock()
```
Should not unlock here!

ianvernon replied:
Good catch, fixed in next upload.

jrajahalme reviewed (approved):
Two unlocks left in the refactored function that defers the unlock now, other than that looks good. I trust you'll take care of these :-)

raybejjani commented on pkg/endpoint/regenerationcontext.go:
```diff
@@ -14,6 +14,13 @@
 
 package endpoint
 
+func (e *ExternalRegenerationMetadata) toRegenerationContext() *regenerationContext {
```
Hmm, since this is now in a separate file from `*ExternalRegenerationMetadata`, maybe it is better as a constructor for `regenerationContext`? This might put us back in the place where we miss parameters again, though :/

ianvernon replied:
I can move `ExternalRegenerationMetadata` to this file, maybe? That would be more organized. I wanted to make the conversion explicit via some function which hides how the conversion is performed.

raybejjani replied:
I actually liked that the thing this function does was here, since `regenerationContext` uses `ExternalRegenerationMetadata`. I had envisioned something like `fromExternalRegenMeta` but the problem there is what type are you getting from this? In contrast `externalRegen.toRegenerationContext` is much clearer. It's just weird to mix them if they aren't in the same file (since by putting them into files we're implying that the file has one function/thing/concept within it).

ianvernon replied:
In master, the code will be in the same file. I'm not sure how the ordering got messed up on GitHub, but both `ExternalRegenerationMetadata` and `regenerationContext` will be in `pkg/endpoint/regenerationcontext.go`.

raybejjani commented on pkg/endpoint/regenerationcontext.go:
```diff
@@ -0,0 +1,34 @@
+
+// ExternalRegenerationMetadata contains any information about a regeneration that
+// the endpoint subsystem should be made aware of for a given endpoint.
+type ExternalRegenerationMetadata struct {
```
Since the code initialises regenerationContext (and datapathRegenerationContext) fields when it needs to anyway. Could the fields in `ExternalRegenerationMetadata` just be public, constant, fields on a single `RegenerationContext`, and the other fields are private and never touched by an external user? I'm not sure it matters that much, in the end, though.

ianvernon replied:
The motivation behind having two separate types is that one is meant to be created outside of this package, while one is created internally (and thus we have more control over it). Although I'm not 100 % sure I understand your proposal - can you provide an example?

raybejjani replied:
Yes, an example might let me prove to myself why what I'm saying makes no sense.
Right now, our code does something like:
- Make a `ExternalRegenerationMetadata` with a reason (e.g. in `daemon.TriggerReloadWithoutCompile`, `daemon.TriggerPolicyUpdate` and `daemon.regenerateRestoredEndpoints`)
- Call `endpointmanager.RegenerateAllEndpoints` or `Endpoint.Regenerate` directly. We end up in `endpoint.Regenerate` anyway.
- `.Regenerate` initialises a `regenerationContext` from the `ExternalRegenerationMetadata`
- various fields are used and set in `regenerationContext` and `datapathRegenerationContext`. E.g. DoneFunc in `Endpoint.Regenerate` [1]

My proposal (including the other one about constructors) was to:
- External caller calls `newRegenerationContext(reason string, reloadDatapath bool)`. This also allocates all fields internally, including `datapathRegenerationContext`.
- Call `endpointmanager.RegenerateAllEndpoints` or `Endpoint.Regenerate` directly. We end up in `endpoint.Regenerate` anyway.
- `.Regenerate` doesn't need to initialise any fields, but can set things like `DoneFunc`

The idea here was that the caller of .Regenerate needed to provide a context, since it was using that package. The package gives it a type that works for this purpose (this is also true of `ExternalRegenerationMetadata`). This single type is ready to go as soon as it is passed to `pkg/endpoint`'s functions.
As you mentioned offline, we initialise things like timeouts and `context.Context` and those are more sensitive to where they are created. This generally makes what I'm saying less appealing, since we would be inconsistent in how some fields are set/created. I did notice that  `datapathRegenerationContext` is what holds these time-sensitive fields, and so it could be created later in the process. This is basically `prepareForProxyUpdates` and so it's acting as a constructor of sorts for this (many fields are set differently).
Anyway, in writing this up I realised how we ended up with these types. I still think we can get away with only `regenerationContext` (but package public) and `datapathRegenerationContext` but this is about 10 lines different from this PR, so not so worth it.


[1] I just noticed that this isn't used outside pkg/endpoint. Can it be private?

ianvernon replied:
The reason I wanted to make two separate types, one external and internal, is that one can be shared by all Endpoints (`ExternalRegenerationMetadata`) vs. one that is local to a single goroutine for a regeneration for a given endpoint `regenerationContext`.  If i follow your proposal, the external caller would allocate the `datapathRegenerationContext`, which I don't want, because I don't want the external caller to be able to instantiate that field because it should only be managed by the goroutine that is regenerating the endpoint.

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -526,234 +526,238 @@ func (e *Endpoint) regenerateBPF(owner Owner, regenContext *regenerationContext)
-	} else {
-		close(datapathRegenCtxt.ctCleaned)
-	}
+	err = func() error {
```
Since this function is so long, a comment explaining what it does and why it's a function-in-a-function would be good (e.g. that it depends on drymode). I'm guessing this is a step towards pulling it all out of here, but if it was all just to ensure the .Unlock is deferred I'm on the fence about it. It's definitely better, but is it too weird a thing to see in the code, especially because it's so long.

raybejjani replied:
Ah, you do factor it out!

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -526,234 +526,238 @@ func (e *Endpoint) regenerateBPF(owner Owner, regenContext *regenerationContext)
+	// reverted in case of failure.
+	// Also keep track of the regeneration finalization code that can't be
+	// reverted, and execute it in case of regeneration success.
+	defer func() {
```
This depends on DryMode, doesn't it? at least, the old code would return when DryMode == true so we never made it to this defer. It also would be skipped in case of errors, so it might need to move below the error return below. In any case, I suspect it is more correct to call this in case of errors, but maybe not in dry mode.

ianvernon replied:
Good catch! I will add a check for dry mode in `finalizeProxyState` in a follow-up commit.

ianvernon replied:
Fixed.

raybejjani reviewed (commented):
I'm +1 on this. My only hangup is that the prepare* functions seem like they could just be one constructor up front. Are the fields reused/initialised in a way that I didn't notice?

Commit:
endpoint: do not finalize proxy state in DryMode

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: remove prepareForDatapathRegeneration function

Instead, setting `ctCleaned` can be done when a `datapathRegenerationContext` is
initialized.

Signed-off by: Ian Vernon <ian@cilium.io>

raybejjani reviewed (approved):
I realised that nothing I said should be blocking so I approved :)
